### ISSUE TITLE
Panic when reporting a result during head-skipping

### ISSUE THREAD

**Issue by charles-paperman** [V0ldek/rsonpath#249]
### Package

Library

### Describe the bug

The following code panick with the attached file (note that a shortest file do not gives the bug)
The error message:
```
thread 'main' panicked at 'range end index 80 out of range for slice of length 64', /home/cpaperma/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:521:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Minimal Reproducible Example

The code generating the error (both function gives the error)

```rust
use rsonpath::engine::{Compiler, Engine, RsonpathEngine};
use rsonpath::input::{OwnedBytes, MmapInput};
use rsonpath::query::JsonPathQuery;
use std::fs::File;

fn loads<'a>(s: &'a str, query_str: &'a str) -> Vec<(usize, usize)>{
    let query = JsonPathQuery::parse(query_str).unwrap();
    let contents = s.to_owned();
    let input = OwnedBytes::try_from(contents).unwrap();
    let engine = RsonpathEngine::compile_query(&query).unwrap();
    let mut results = vec![];
    engine.matches(&input, &mut results).unwrap();
    let cast = results.into_iter().map(|m| (m.span.start_idx, m.span.end_idx) ).collect(); 
    cast
}

unsafe fn load<'a>(path: &'a str, query_str: &'a str) -> Vec<String> {
    let query = JsonPathQuery::parse(query_str).unwrap();
    let file = File::open(path).unwrap();
    let input = MmapInput::map_file(&file).unwrap(); 
    let engine = RsonpathEngine::compile_query(&query).unwrap();
    let mut results = vec![];
    engine.matches(&input, &mut results).unwrap();
    let cast = results.into_iter().map(|m| String::from_utf8(m.bytes).unwrap() ).collect(); 
    cast
}

fn main() {
    unsafe{
        let it = load("0.json", "..author");
        println!("{}", it[0]);
    }
}
```

The file:
[0.json.gz](https://github.com/V0ldek/rsonpath/files/12538762/0.json.gz)

### Expected behavior

No bug

### Version of the release

v0.7.0

### Rust version

rustc 1.70.0 (90c541806 2023-05-31)

### Target triple

x86_64

### Features enabled

default

**Comment by V0ldek** [comment]
Confirmed repro, here's the backtrace:

```ini
thread 'main' panicked at 'range end index 80 out of range for slice of length 64', /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:521:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
   2: core::slice::index::slice_end_index_len_fail_rt
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/slice/index.rs:76:5
   3: core::slice::index::slice_end_index_len_fail
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/slice/index.rs:68:9
   4: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/slice/index.rs:408:13
   5: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/slice/index.rs:18:9
   6: rsonpath::result::nodes::append_final_block
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:521:18
   7: rsonpath::result::nodes::SimpleRecorder<B,S>::record_value_terminator
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:291:17
   8: rsonpath::result::nodes::InternalRecorder<B,S>::record_value_terminator
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:236:32
   9: <rsonpath::result::nodes::NodesRecorder<B,S> as rsonpath::result::Recorder<B>>::record_value_terminator
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/result/nodes.rs:109:9
  10: rsonpath::engine::head_skipping::HeadSkip<I,_>::run_head_skipping
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/engine/head_skipping.rs:218:48
  11: rsonpath::engine::main::Executor<I,R>::run
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/engine/main.rs:206:32
  12: <rsonpath::engine::main::MainEngine as rsonpath::engine::Engine>::matches
             at /home/mat/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.7.0/src/engine/main.rs:110:9
  13: rsonpath_249_repro::load
             at ./src/main.rs:23:5
  14: rsonpath_249_repro::main
             at ./src/main.rs:30:18
  15: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by V0ldek** [comment]
It doesn't happen with `simd` disabled, and it doesn't happen if you use `$[*]..author` as the query, which returns the same results. So the error is most likely in head skipping.

**Comment by V0ldek** [comment]
This also happens in the binary if you run `rq '$..author' 0.json` on v0.7.0.

**Comment by V0ldek** [comment]
One way to reproduce this with a small document is to have `memmem` look for a very long label. Full repro:

```json
{
  "very long label to search for, like, extremely long, so that the colon occurs really far away from the start of the needle match, which triggers some interesting behavior and might break the head skipping module like in #249": 42
}
```

query: `$..['very long label to search for, like, extremely long, so that the colon occurs really far away from the start of the needle match, which triggers some interesting behavior and might break the head skipping module like in #249']`

Expected result is the node `42`, span `232-233`. The panic is the same path as in the original bug report:

```ini
panicked at 'range end index 172 out of range for slice of length 64', /home/mat/rsonpath/crates/rsonpath-lib/src/result/nodes.rs:521:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/slice/index.rs:76:5
   3: core::slice::index::slice_end_index_len_fail
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/slice/index.rs:68:9
   4: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/slice/index.rs:411:13
   5: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/slice/index.rs:18:9
   6: rsonpath::result::nodes::append_final_block
             at ./src/result/nodes.rs:521:18
   7: rsonpath::result::nodes::SimpleRecorder<B,S>::record_value_terminator
             at ./src/result/nodes.rs:291:17
   8: rsonpath::result::nodes::InternalRecorder<B,S>::record_value_terminator
             at ./src/result/nodes.rs:236:32
   9: <rsonpath::result::nodes::NodesRecorder<B,S> as rsonpath::result::Recorder<B>>::record_value_terminator
             at ./src/result/nodes.rs:109:9
  10: rsonpath::engine::head_skipping::HeadSkip<I,_>::run_head_skipping
             at ./src/engine/head_skipping.rs:218:48
  11: rsonpath::engine::main::Executor<I,R>::run
             at ./src/engine/main.rs:206:32
  12: <rsonpath::engine::main::MainEngine as rsonpath::engine::Engine>::matches
             at ./src/engine/main.rs:110:9
  13: end_to_end::extremely_long_label_to_search_with_query_select_the_extremely_long_label_with_buffered_input_and_nodes_result_using_main_engine
             at ./tests/end_to_end.rs:8296:5
  14: end_to_end::extremely_long_label_to_search_with_query_select_the_extremely_long_label_with_buffered_input_and_nodes_result_using_main_engine::{{closure}}
             at ./tests/end_to_end.rs:8289:6
  15: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
  16: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
```

I think there might be an even edgier edge case to find here, though.

**Comment by V0ldek** [comment]
I have an even more insane test case.

```json
{
  "target": {
    "please note the important whitespaces after the upcoming comma (pretend indentation is really big)": 42
  },                                                   
  "target" : 43
}
```

with query `$..target` also crashes on the same path.

**Comment by V0ldek** [comment]
This will be fixed in the next release. Will have to measure performance impact, but it should be positive as I managed to make the head-skipping logic a bit simpler.

**Comment by charles-paperman** [comment]
Congrats for the bug fixes!

**Comment by V0ldek** [comment]
Confirmed fixed and released in v0.7.1
